synapse/lib/lmdbslab.py

```diff
@@ -363,8 +393,109 @@ def _growMapSize(self, size=None):
+                'too low for optimal performance.', locked_ulimit)
+
+        logger.debug('memory locking thread started')
+
+        # Note:  available might be larger than max_total in a container
+        max_total = s_thisplat.getTotalMemory()
+        available = s_thisplat.getAvailableMemory()
+
+        PAGESIZE = 4096
+        max_to_lock = (min(locked_ulimit,
+                           int(max_total * MAX_TOTAL_PERCENT),
+                           int(available * MAX_TOTAL_PERCENT)) // PAGESIZE) * PAGESIZE
+
+        self.max_could_lock = max_to_lock
+
+        path = self.path.absolute() / 'data.mdb'  # Path to the file that gets mapped
+        fh = open(path, 'r+b')
+        fileno = fh.fileno()
+
+        prev_memstart = 0  # The last start of file mapping, used to keep track when the mapping moves
+        prev_memend = 0  # The last end of the file mapping, so we can start from there
+
+        # Avoid spamming messages
+        first_end = True
+        limit_warned = False
+        self.locking_memory = True
+
+        self.resizeevent.set()
+
+        while not self.isfini:
+
+            self.resizeevent.wait()
+            if self.isfini:
+                break
+
+            self.resizeevent.clear()
+
+            memstart, memlen = s_thisplat.getFileMappedRegion(path)
+            if memlen > max_to_lock:
+                memlen = max_to_lock
+                if not limit_warned:
+                    logger.warning('memory locking limit reached')
+                    limit_warned = True
+                # Even in the event that we've hit our limit we still have to loop because further mmaps may cause
+                # the base address to change, necessitating relocking what we can
+
+            # The file might be a little bit smaller than the map because rounding (and mmap fails if you give it a
+            # too-long length)
+            filesize = os.fstat(fileno).st_size
+            goal_end = memstart + min(memlen, filesize)
+            self.lock_goal = goal_end
+
+            # If we just started or we evidently got a new base address, restart mapping from the beginning
+            if not prev_memstart or prev_memstart != memstart:
+                self.lock_progress = 0
+                prev_memend = prev_memstart = memstart
+
+            assert goal_end >= prev_memend
+
+            # Actually do the prefaulting and locking.  Only do it a chunk at a time to maintain responsiveness.
+            while prev_memend < goal_end:
+                new_memend = min(prev_memend + MAX_LOCK_AT_ONCE, goal_end)
+                memlen = new_memend - prev_memend
+                PROT = 1 # PROT_READ
+                FLAGS = 0x8001  # MAP_POPULATE | MAP_SHARED (Linux only)  (for fast prefaulting)
+                try:
+                    self.prefaulting = True
+                    with s_thisplat.mmap(0, length=new_memend - prev_memend, prot=PROT, flags=FLAGS, fd=fileno,
```

Comment:
Maybe we should create one big memory map at once to avoid creating a zillion separate maps and then lock them incrementally?

Reply:
Then you won't be able to kill it. This might take minutes on a large cortex The chunks are already pretty big (a GiB), and we can always make them bigger.